README.md
# @goodware/task-queue: A lightweight task queue

## Links

- [Release History](https://good-ware.github.io/js-task-queue/tutorial-Releases)
- [npm](https://www.npmjs.com/package/@goodware/task-queue)
- [git](https://github.com/good-ware/js-task-queue)
- [API](https://good-ware.github.io/js-task-queue/)

## Requirements

NodeJS 8+

## Installation

`npm i --save @goodware/task-queue`

## Overview

This lightweight, battle-tested, single-dependency (Joi 17) task queue limits the number of tasks (sychronous or asynchronous) that execute concurrently. The purpose of limiting task execution is to control resource usage such as memory and database connections.

Although several packages address this use case, this is apparently the only library that can queue tasks post-instantiation without using generators.

## Creation

A task-queue object is instantiated by providing a configuration object to the constructor. The configuration object currently has one required and one optional property:

| Name      | Description                                                              |
| --------- | ------------------------------------------------------------------------ |
| `size`    | The size of the queue (required)                                         |
| `workers` | The number of tasks that can execute simultaneously. Defaults to `size.` |

## Usage

Functions are queued via the asynchronous method `push(task)`. This method accepts a function named `task` and returns a Promise that resolves to an object when the `task` function is called (_not_ when `task` returns). `task` is called only when a worker is available. `task` does not need to return a Promise, but if it does, it can be acquired via the `promise` property of the object returned by `push().`

### Brief Example

- Create a queue that runs at most 10 running tasks
```js
new (require('@goodware/task-queue'))({ size: 10 });
```
- Wait for the provided function to be invoked: 
```js
await queue.push(() => {...})
```
- Wait for the provided function to finish:
```js
await (await queue.push(() => {...})).promise;
```

### Code Sample

This example runs at most two tasks at a time. It outputs: 2, 1, 4, 3.

```js
const queue = new (require('@goodware/task-queue'))({ size: 2 });

// Task #1 : await push() returns immediately because the queue is empty. 'await'
// doesn't wait for the task to complete.
await queue.push(
  () =>
    new Promise((resolve) =>
      setTimeout(() => {
        console.log(`Task 1 ${Date.now()}`);
        resolve();
      }, 400)
    )
);

// Task #2 : await push() returns immediately because the queue has an open slot
await queue.push(
  () =>
    new Promise((resolve) =>
      setTimeout(() => {
        console.log(`Task 2 ${Date.now()}`);
        resolve();
      }, 300)
    )
);

// The queue is full. Task #2 will finish in about 300 ms.

// Task #3 : await push() waits until task #2 finishes
await queue.push(
  () =>
    new Promise((resolve) =>
      setTimeout(() => {
        console.log(`Task 3 ${Date.now()}`);
        resolve();
      }, 200)
    )
);

// The queue is full again. 300 ms have already passed. Task #1 will
// terminate in about 100 ms, leaving task #3 in the queue.

// Task #4 : await push() waits until task #1 finishes
const ret = await queue.push(
  () =>
    new Promise((resolve) =>
      setTimeout(() => {
        console.log(`Task 4 ${Date.now()}`);
        resolve();
      }, 100)
    )
);

// Wait for task #4 to finish
await ret.promise;
```

## Minimizing Memory Usage

`push()` returns a new Promise each time it is called, thus consuming memory. Depending on your application, it may be necessary to limit calls to `push()` when the queue is full if you are unable to control the number of calls to `push().`

For example, consider the following constraints:

1. Up to 10 workers can execute at the same time
2. When 10 workers are running, up to 50 tasks can call `push()` and immediately continue their work. Subsequent callers will wait until a worker has finished.

Although it appears that resources are properly constrained in this scenario, if `push()` is called, say, 1,000 times a second, and the workers take longer than 1 second each, the process will likely run out of memory. One solution to this scenario is backpressure.

No form of backpressure is a silver bullet. External systems must handle errors and retry.

### Code Sample

```js
const queue = new (require('@goodware/task-queue'))({ size: 50, workers: 10 });

for (;;) {
  // ...
  // The most basic implementation of backpressure
  if (!queue.full) {
    await workers.push(async () => {
      await doWork();
    });
  }
  // ...
}

await queue.stop();
```


moredocs/Releases.md
## 2.0.4 2021-11-07

Fix [#16](https://github.com/good-ware/js-task-queue/issues/16)

## 2.0.3 2021-06-07

Update docs

## 2.0.2 2021-05-03

Update docs

## 2.0.1 2021-05-03

Documentation and minor bug fixes

## 2.0.0 2021-05-03

### Breaking Changes

- full() and isFull() replaced by full
- pushUnlessFull() removed

## 1.2.0 2021-05-02

### New Features

- New option `workers.` Defaults to `size.` Consider `size=3` and `workers=1.` Lines 1-3 in the following example would return immediately even though only one function executes at a time. Line 4, however, would wait until one of the tasks started by lines 1-3 complete.

```js
await queue.push(() => a); // Line 1
await queue.push(() => b); // Line 2
await queue.push(() => c); // Line 3
await queue.push(() => c); // Line 4
```


tests/test.js
/* eslint-disable no-console */
const TaskQueue = require('../index');

/**
 * @description Test
 * @return {Promise}
 */
async function test1() {
  // The following code will only run two promises concurrently at once.
  const queue = new TaskQueue({ size: 2 });
  const ret = await queue.push(() => new Promise((resolve) => setTimeout(() => resolve(new Date()), 3000)));
  console.log(`Waiting for value ${new Date()}`);
  console.log(`Done waiting for value: ${await ret.promise}`);
}

/**
 * @description Test
 * @return {Promise}
 */
async function test2() {
  // The following code will only run two promises concurrently at once.
  const queue = new TaskQueue({ size: 2 });
  const ret = await queue.push(
    () => new Promise((resolve, reject) => setTimeout(() => reject(new Error('hello there')), 2000))
  );
  console.log(`Waiting for error ${new Date()}`);

  let failed;

  try {
    await ret.promise;
    failed = true;
  } catch (error) {
    console.log(`Done waiting for error. ${new Date()} ${error.message}`);
  }

  if (failed) throw new Error('Failed to throw exception');
}

/**
 * @description Test
 * @return {Promise}
 */
async function test3() {
  const queue = new TaskQueue({ size: 2 });
  await queue.push(() => new Promise((resolve) => setTimeout(() => resolve(), 100)));
  await queue.wait();
}

/**
 * @description Test
 * @return {Promise}
 */
async function test4() {
  const queue = new TaskQueue({ size: 2 });
  await queue.push(() => new Promise((resolve) => setTimeout(() => resolve(), 100)));
  await queue.stop();
}

/**
 * @description Runs all tests
 * @return {Promise}
 */
function go() {
  return Promise.all([test1(), test2(), test3(), test4()]);
}

go().then(() => console.log('Successful'), console.error);


tests/index.js
/* eslint-disable no-console */
const TaskQueue = require('../index');

/**
 * @description Test
 * @return {Promise}
 */
async function test1() {
  // The following code will only run two promises concurrently at once.
  const queue = new TaskQueue({ size: 2 });
  const ret = await queue.push(() => new Promise((resolve) => setTimeout(() => resolve(new Date()), 3000)));
  console.log(`Waiting for value ${new Date()}`);
  console.log(`Done waiting for value: ${await ret.promise}`);
}

/**
 * @description Test
 * @return {Promise}
 */
async function test2() {
  // The following code will only run two promises concurrently at once.
  const queue = new TaskQueue({ size: 2 });
  const ret = await queue.push(
    () => new Promise((resolve, reject) => setTimeout(() => reject(new Error('hello there')), 2000))
  );
  console.log(`Waiting for error ${new Date()}`);

  let failed;

  try {
    await ret.promise;
    failed = true;
  } catch (error) {
    console.log(`Done waiting for error. ${new Date()} ${error.message}`);
  }

  if (failed) throw new Error('Failed to throw exception');
}

/**
 * @description Test
 * @return {Promise}
 */
async function test3() {
  const queue = new TaskQueue({ size: 2 });
  await queue.push(() => new Promise((resolve) => setTimeout(resolve, 100)));
  await queue.wait();
}

/**
 * @description Test
 * @return {Promise}
 */
async function test4() {
  const queue = new TaskQueue({ size: 2 });
  await queue.push(() => new Promise((resolve) => setTimeout(resolve, 100)));
  await queue.stop();
}

/**
 * @description Test workers < size
 * @return {Promise}
 */
async function test5() {
  console.log(`test5> ${new Date()}`);
  const queue = new TaskQueue({ size: 3, workers: 1 });

  await queue.push(() => new Promise((resolve) => setTimeout(resolve, 2000)));
  const ret = await queue.push(
    () =>
      new Promise((resolve) =>
        setTimeout(() => {
          resolve(5);
        }, 2000)
      )
  );
  console.log(`test5> I run immediately ${new Date()}`);
  await queue.push(() => new Promise((resolve) => setTimeout(resolve, 2000)));
  console.log(`test5> I run immediately ${new Date()}`);

  ret.promise.then((value) => {
    if (value !== 5) throw new Error('failed');
    console.log(`test5> 4 seconds later ... ${value} === 5 ${new Date()}`);
  });

  const ret2 = await queue.push(() => new Promise((resolve) => setTimeout(resolve, 1000)));
  console.log(`test5> 6 seconds later ${new Date()}`);

  ret2.promise.then(() => console.log(`test5> 7 seconds later ${new Date()}`));

  await queue.push(() => new Promise((resolve) => setTimeout(resolve, 1000)));
  await queue.stop();
  console.log(`test5> 8 seconds later ${new Date()}`);
}

/**
 * @description Test full
 * @return {Promise}
 */
async function test6() {
  const queue = new TaskQueue({ size: 2, workers: 1 });
  await queue.push(() => new Promise((resolve) => setTimeout(resolve, 100)));
  if (queue.full) throw new Error('failed');
  await queue.push(() => new Promise((resolve) => setTimeout(resolve, 100)));
  if (!queue.full) throw new Error('failed');
  await queue.stop();
}

/**
 * @description Runs all tests
 * @return {Promise}
 */
function go() {
  return Promise.all([test1(), test2(), test3(), test4(), test5(), test6()]);
}

go().then(
  () => console.log('Successful'),
  (error) => {
    console.error(`Failed: ${error}`);
  }
);


tests/example.js
/* eslint-disable no-console */

// This script outputs (roughly): Task 2, Task 1, Task 4, Task 3

(async () => {
  // eslint-disable-next-line global-require, import/no-unresolved
  const queue = new (require('@goodware/task-queue'))({ size: 2 });

  // Task #1 : push() returns immediately because the queue is empty. 'await' doesn't wait for the task to complete.
  await queue.push(
    () =>
      new Promise((resolve) =>
        setTimeout(() => {
          console.log(`Task 1 ${Date.now()}`);
          resolve();
        }, 400)
      )
  );

  // Task #2 : push() returns immediately because the queue has an open slot
  await queue.push(
    () =>
      new Promise((resolve) =>
        setTimeout(() => {
          console.log(`Task 2 ${Date.now()}`);
          resolve();
        }, 300)
      )
  );

  // The queue is full. Task #2 will finish in about 300 ms.

  // Task #3 : push() waits until task #2 finishes
  await queue.push(
    () =>
      new Promise((resolve) =>
        setTimeout(() => {
          console.log(`Task 3 ${Date.now()}`);
          resolve();
        }, 200)
      )
  );

  // The queue is full again. 300 ms have already passed. Task #1 will terminate in about 100 ms, leaving task #3 in the
  // queue.

  // Task #4 : push() waits until task #1 finishes
  const ret = await queue.push(
    () =>
      new Promise((resolve) =>
        setTimeout(() => {
          console.log(`Task 4 ${Date.now()}`);
          resolve();
        }, 100)
      )
  );

  // Wait for task #4 to finish
  await ret.promise;
})();


index.js
/* eslint-disable no-await-in-loop */
/* eslint-disable no-plusplus */
const Joi = require('joi');

/**
 * @description Schema for TaskQueue's constructor options
 */
const optionsSchema = Joi.object({
  name: Joi.string().description('The name of the queue. It is logged.'),
  size: Joi.number()
    .integer()
    .min(1)
    .required()
    .description(`push() resolves when the task queue's length is less than this value`),
  workers: Joi.number()
    .integer()
    .min(1)
    .description('The maximum number of tasks that can execute simultaneously. Defaults to size.'),
  logger: Joi.object(),
});

const logTag = 'taskQueue';

/**
 * @description A queue that enforces a maximum number of simultaneously executing tasks
 * @todo Add events
 */
class TaskQueue {
  /**
   * Properties:
   *  {Boolean} stopping
   *  {Boolean} stopped
   *  {Object} logger
   *  {Integer} taskCount The number of currently executing tasks
   *  {Function[]} doneListeners
   *  {Function[]} waitListeners
   */
  /**
   * @description Constructor. There is no need to call start() after creating a new object.
   * @param {Object} options
   * @param {Integer} options.size The maximum number of functions that can execute at a time
   */
  constructor(options) {
    const validation = optionsSchema.validate(options);
    if (validation.error) throw new Error(validation.error.message);

    Object.assign(this, validation.value);

    if (!this.workers) this.workers = this.size;

    if (this.logger && !this.logger.isLevelEnabled(logTag)) delete this.logger;

    this.taskCount = 0;
    this.doneListeners = [];
    this.waitListeners = [];
  }

  /**
   * @private
   * @ignore
   * @description Called when a task finishes
   */
  taskFinished() {
    const newTasks = this.taskCount - 1;

    if (this.logger) {
      this.logger.log(logTag, {
        message: `Task finished for '${this.name}'. Tasks: ${newTasks}`,
        name: this.name,
        taskCount: newTasks,
      });
    }

    if (newTasks < 0) {
      const message = `Task counter is negative for '${this.name}'`;
      if (this.logger) {
        this.logger.log(['error', logTag], {
          message,
          name: this.name,
        });
      }
    } else {
      this.taskCount = newTasks;
    }

    // =======================
    // Fire an event to done()
    // This can cause reentrancy
    if (this.doneListeners.length) this.doneListeners.shift()();

    // =======================
    // Fire an event to wait()
    if (!this.taskCount) {
      const { waitListeners } = this;
      if (waitListeners.length) {
        this.waitListeners = [];
        waitListeners.forEach((resolve) => resolve());
      }
    }
  }

  /**
   * @description Starts a task. If the queue's maximum size has been reached, this method waits for a task to finish
   *  before invoking task().
   * @param {Function} task A function to call. It can return a Promise, throw an exception, or return a value.
   * @return {Promise} Does not reject. Resolves to an object with the property 'promise'
   *  containing either the Promise returned by task or a new Promise that resolves to the value returned by task or
   *  rejects using the exception thrown by it. Therefore, it is not only possible to wait for the task to start, it is
   *  also possible to wait for it to finish.
   *
   *  For example:
   *  // Wait for an open slot in the queue
   *  const ret = await queue.push(()=>new Promise(resolve=>setTimeout(()=>resolve('Hello'), 5000)));
   *  // Wait for 5 seconds and output Hello
   *  console.log(await ret.promise);
   */
  async push(task) {
    // Wait for an available slot in the queue
    // eslint-disable-next-line no-await-in-loop
    for (;;) {
      if (this.taskCount < this.workers) break;
      if (this.full) {
        await new Promise((resolve) => this.doneListeners.push(resolve));
      } else {
        // =================
        // The queue is full
        const promise = new Promise((resolve, reject) =>
          this.doneListeners.push(() =>
            this.push(task).then((ret) => {
              // ret.promise is from the task. Forward to the Promise returned by this method.
              ret.promise.then(resolve, reject);
            })
          )
        );
        return { promise };
      }
    }

    let fret;
    let err;

    try {
      fret = task(); // this could throw an exception if it's not explicitly marked async, so increment
    } catch (error) {
      err = error;
    }

    // Increment taskCount here, not earlier
    const taskCount = ++this.taskCount;

    if (this.logger) {
      this.logger.log(logTag, {
        message: `Task started for '${this.name}'. Tasks: ${taskCount}`,
        name: this.name,
        taskCount,
      });
    }

    let promise;

    if (fret && Object.prototype.toString.call(fret) === "[object Promise]") {
      promise = new Promise((resolve, reject) => {
        fret.then(
          (value) => {
            this.taskFinished();
            resolve(value);
          },
          (error) => {
            this.taskFinished();
            reject(error);
          }
        );
      });
    } else {
      promise = err ? Promise.reject(err) : Promise.resolve(fret);
      this.taskFinished();
    }

    // If bare 'promise' was returned, the caller would wait for fret to resolve. Instead, callers should only wait for
    // for task to be called. Therefore, return an object that contains a 'promise' key. Awaiting on this method
    // therefore waits for an empty slot in the queue and returns a Promise that immediately resolves to an object.
    return { promise };
  }

  /**
   * @description Is the queue full?
   * @return {Boolean} true if the maximum number of tasks are queued
   */
  get full() {
    return this.taskCount + this.doneListeners.length >= this.size;
  }

  /**
   * @description Waits for running tasks to complete. Callers are not prevented callers from calling push(); thus
   *  there is no guarantee that when the returned Promise resolves, the queue will have an available slot.
   * @return {Promise}
   */
  async wait() {
    // eslint-disable-next-line no-await-in-loop
    if (this.taskCount) await new Promise((resolve) => this.waitListeners.push(resolve));
  }

  /**
   * @description Waits for running tasks to complete. Prevents additional calls to push().
   */
  async stop() {
    this.stopping = true;
    await this.wait();
    this.stopping = false;
    this.stopped = true;
  }

  /**
   * @description Undo method for stop(). There is no need to invoke start() after creating a new object.
   */
  start() {
    if (!this.stopped) return;
    if (this.stopping) throw new Error('Stopping');
    this.stopped = false;
  }
}

module.exports = TaskQueue;


TaskQueue.js
/* eslint-disable no-plusplus */
const Joi = require('joi');

const optionsSchema = Joi.object({
  name: Joi.string().description('The name of the queue. It is logged.'),
  size: Joi.number()
    .integer()
    .min(1)
    .required()
    .description('The maximum number of simultaneous tasks that can execute'),
  logger: Joi.object(),
});

const logTag = 'taskQueue';

/**
 * @description A queue of executing tasks with a maximum size
 * @todo Add events
 */
class TaskQueue {
  /**
   * Properties:
   *  {Boolean} stopping
   *  {Boolean} stopped
   *  {Object} logger
   *  {Number} taskCount The number of currently executing tasks
   *  {Function[]} resolvers
   */
  /**
   * @description Constructor. There is no need to call start() after creating a new object.
   * @param {Object} options
   */
  constructor(options) {
    const validation = optionsSchema.validate(options);
    if (validation.error) throw new Error(validation.error.message);

    Object.assign(this, validation.value);

    if (this.logger && !this.logger.isLevelEnabled(logTag)) delete this.logger;

    this.taskCount = 0;
    this.resolvers = [];
  }

  /**
   * @description Called when a task finishes
   * @private
   */
  taskFinished() {
    const newTasks = this.taskCount - 1;

    if (this.logger) {
      this.logger.log(logTag, {
        message: `Task finished for '${this.name}'. Tasks: ${newTasks}`,
        name: this.name,
        taskCount: newTasks,
      });
    }

    if (newTasks < 0) {
      const message = `Task counter is negative for '${this.name}'`;
      if (this.logger) {
        this.logger.log(['error', logTag], {
          message,
          name: this.name,
        });
      }
    } else {
      this.taskCount = newTasks;
    }

    // Invoke resolve() for all waiters. This is a while loop because of wait()
    while (this.resolvers.length) this.resolvers.shift()();
  }

  /**
   * @description Starts a task. If the queue's maximum size has been reached, this method waits for a task to finish
   *  before invoking task().
   * @param {Function} task A function to call. It can return a Promise, throw an exception, or return a value.
   * @return {Promise} Resolves to an object with the property 'promise' containing either the Promise returned by task
   *  or a new Promise that resolves to the value returned by task or rejects using the exception thrown by it.
   *  Therefore, it is not only possible to wait for the task to start, it is also possible to wait for it to finish.
   *  For example:
   *  // Wait for an open slot in the queue
   *  const ret = await queue.push(()=>new Promise(resolve=>setTimeout(()=>resolve('Hello'), 5000)));
   *  // Wait for 5 seconds and output Hello
   *  console.log(await ret.promise);
   */
  async push(task) {
    // Wait for an available slot in the queue
    // eslint-disable-next-line no-await-in-loop
    while (this.taskCount >= this.size) await new Promise((resolve) => this.resolvers.push(resolve));

    let fret;
    let err;

    try {
      fret = task(); // this could throw an exception if it's not explicitly marked async, so increment
    } catch (error) {
      err = error;
    }

    // Increment taskCount here, not earlier
    const taskCount = ++this.taskCount;

    if (this.logger) {
      this.logger.log(logTag, {
        message: `Task started for '${this.name}'. Tasks: ${taskCount}`,
        name: this.name,
        taskCount,
      });
    }

    let promise;

    if (fret instanceof Promise) {
      promise = new Promise((resolve, reject) => {
        fret.then(
          (value) => {
            this.taskFinished();
            resolve(value);
          },
          (error) => {
            this.taskFinished();
            reject(error);
          }
        );
      });
    } else {
      promise = err ? Promise.reject(err) : Promise.resolve(fret);
      this.taskFinished();
    }

    // If bare 'promise' was returned, the caller would wait for fret to resolve. Instead, callers should only wait for
    // for task to be called. Therefore, return an object that contains a 'promise' key. Awaiting on this method
    // therefore waits for an empty slot in the queue and returns a Promise that immediately resolves to an object.
    return { promise };
  }

  /**
   * @description Is the queue full?
   * @return {Boolean} true if the maximum number of tasks are queued
   */
  isFull() {
    return this.taskCount >= this.size;
  }

  /**
   * @description Calls push(task) if the queue has an available slot. If the queue is full, this method returns
   *  immediately with a falsey value. Otherwise, it returns the same value as the push() method.
   * @param {Function} task
   * @return {Promise}
   */
  pushIfAvailable(task) {
    if (this.taskCount >= this.size) return false;
    return this.push(task);
  }

  /**
   * @description Waits for running tasks to complete. Callers are not prevented callers from calling push(); thus
   *  there is no guarantee that when the returned Promise resolves, the queue will actually be empty.
   * @return {Promise}
   */
  async wait() {
    // eslint-disable-next-line no-await-in-loop
    while (this.taskCount) await new Promise((resolve) => this.resolvers.push(resolve));
  }

  /**
   * @description Waits for running tasks to complete. Prevents additional calls to push().
   */
  async stop() {
    this.stopping = true;
    await this.wait();
    this.stopping = false;
    this.stopped = true;
  }

  /**
   * @description Undo method for stop(). There is no need to invoke start() after creating a new object.
   */
  start() {
    if (!this.stopped) return;
    if (this.stopping) throw new Error('Stopping');
    this.stopped = false;
  }
}

module.exports = TaskQueue;


AsyncQueue.js
/* eslint-disable no-plusplus */
const Joi = require('joi');

const optionsSchema = Joi.object({
  name: Joi.string().description('(Optional) The name of this scheduler. It is logged.'),
  size: Joi.number()
    .integer()
    .min(1)
    .required()
    .description('The maximum number of simultaneous tasks that can execute'),
  logger: Joi.object(),
});

/**
 * @description Promise queue/scheduler. Manages tasks to execute in a queue with a maximum size.
 * @example
 *   // The following script instantiates at most two promises at a time. It outputs roughly: 2, 1, 4, 3
 *   const scheduler = new Scheduler({size: 2});
 *   await scheduler.push(()=>new Promise(resolve=>setTimeout(()=>resolve(console.log('1 '+Date.now())), 400)));
 *   await scheduler.push(()=>new Promise(resolve=>setTimeout(()=>resolve(console.log('2 '+Date.now())), 300)));
 *   // The next line waits a little more than 300ms for the line above it to finish
 *   await scheduler.push(()=>new Promise(resolve=>setTimeout(()=>resolve(console.log('3 '+Date.now())), 200)));
 *   await scheduler.push(()=>new Promise(resolve=>setTimeout(()=>resolve(console.log('4 '+Date.now())), 100)));
 *
 * Properties:
 *  {Boolean} stopping
 *  {Boolean} stopped
 *  {Object} logger
 *  {Number} taskCount The number of currently executing tasks
 *  {Array} waiters
 */
class Scheduler {
  /**
   * @description Constructor. There is no need to invoke start() after creating a new object.
   * @param {Object} options
   */
  constructor(options) {
    const validation = optionsSchema.validate(options);
    if (validation.error) throw new Error(validation.error.message);

    Object.assign(this, validation.value);

    // This class only logs at the default level
    if (this.logger && this.logger.isLevelEnabled()) this.logger = this.logger.logger('scheduler');

    this.taskCount = 0;
    this.waiters = [];
  }

  /**
   * @description Called when a task finishes
   */
  taskFinished() {
    const newTasks = this.taskCount - 1;

    if (this.logger) {
      this.logger.default({
        message: `Task finished for '${this.name}'. Tasks: ${newTasks}`,
        name: this.name,
        taskCount: newTasks,
      });
    }

    if (newTasks < 0) {
      const message = `Task counter is negative for '${this.name}'`;
      if (this.logger) {
        this.logger.error({
          message,
          name: this.name,
        });
      }
    } else {
      this.taskCount = newTasks;
    }

    // This is a while loop because of wait()
    while (this.waiters.length) {
      const waiter = this.waiters.shift();
      waiter();
    }
  }

  /**
   * @description Starts a task. If the queue's maximum size has been reached, this method waits for a task to finish
   *  before calling func().
   * @param {Function} func Function to call. It can return a Promise, throw an exception, or return a value.
   * @return {Promise} Resolves to an object with the key 'promise' containing either the Promise returned by func or
   *  a new Promise that resolves to the value returned by func or rejects using the exception thrown by it. Therefore,
   *  it is not only possible to wait for func to start, it is also possible to wait for it to finish. For example:
   *  // Wait for an open slot in the queue
   *  const ret = await scheduler.push(()=>new Promise(resolve=>setTimeout(()=>resolve('Hello'), 5000)));
   *  // Wait for 5 seconds and output Hello
   *  console.log(await ret.promise);
   */
  async push(func) {
    // Wait for an available slot in the queue
    // eslint-disable-next-line no-await-in-loop
    while (this.taskCount >= this.size) await new Promise((resolve) => this.waiters.push(resolve));

    let fret;
    let err;

    try {
      fret = func(); // this could throw an exception if it's not explicitly marked async, so increment
    } catch (error) {
      err = error;
    }

    // Increment taskCount here, not earlier
    const taskCount = ++this.taskCount;

    if (this.logger) {
      this.logger.default({
        message: `Task started for '${this.name}'. Tasks: ${taskCount}`,
        name: this.name,
        taskCount,
      });
    }

    let promise;

    if (fret instanceof Promise) {
      promise = new Promise((resolve, reject) => {
        fret.then(
          (value) => {
            this.taskFinished();
            resolve(value);
          },
          (error) => {
            this.taskFinished();
            reject(error);
          }
        );
      });
    } else {
      promise = err ? Promise.reject(err) : Promise.resolve(fret);
      this.taskFinished();
    }

    // If bare 'promise' was returned, the caller would wait for fret to resolve. Instead, callers should only wait for
    // for func to be called. Therefore, return an object that contains a 'promise' key. Awaiting on this method
    // therefore waits for an empty slot in the queue and returns a Promise that immediately resolves to an object.
    return { promise };
  }

  /**
   * @description Returns true if the maximum number of tasks are running
   * @return {Boolean}
   */
  full() {
    return this.taskCount >= this.size;
  }

  /**
   * @description Calls push(func) if the queue has an available slot
   * @param {Function} func
   * @return {Promise} Returns a falsey value if full
   */
  pushIfAvailable(func) {
    if (this.taskCount >= this.size) return false;
    return this.push(func);
  }

  /**
   * @description Waits for all running tasks to complete. Callers are not prevented callers from calling push(); thus
   *  there is no guarantee that when the returned Promise resolves, the queue will actually be empty.
   * @return {Promise}
   */
  async wait() {
    // eslint-disable-next-line no-await-in-loop
    while (this.taskCount) await new Promise((resolve) => this.waiters.push(resolve));
  }

  /**
   * @description Waits for all running tasks to complete. Prevents additional calls to push().
   */
  async stop() {
    this.stopping = true;
    await this.wait();
    this.stopping = false;
    this.stopped = true;
  }

  /**
   * @description Undo method for stop(). There is no need to invoke start() after creating a new object.
   */
  start() {
    if (!this.stopped) return;
    if (this.stopping) throw new Error('Stopping');
    this.stopped = false;
  }
}

module.exports = Scheduler;
